src/payment/strategies/braintree/braintree-credit-card-payment-strategy.ts

```diff
@@ -177,6 +191,35 @@ export default class BraintreeCreditCardPaymentStrategy implements PaymentStrate
         };
     }
 
+    private async _processAdditionalAction(
+        error: Error,
+        payment: OrderPaymentRequestBody,
+        billingAddress: Address,
+        orderAmount: number
+    ): Promise<InternalCheckoutSelectors> {
+        // test: check handlerError function
+        if (!(error instanceof RequestError) || !some(error.body.errors, { code: 'three_d_secure_required' })) {
+            return this._handleError(error);
+        }
+
+        try {
+            const { payer_auth_request: storedCreditCardNonce } = error.body.three_ds_result || {};
+
+            return await this._store.dispatch(this._paymentActionCreator.submitPayment({
+                ...payment,
+                paymentData: {
+                    deviceSessionId: this._deviceSessionId,
+                    ...this._mapToNonceInstrument({
+                        ...payment.paymentData,
+                        ...await this._braintreePaymentProcessor.challenge3DSVerification(storedCreditCardNonce, orderAmount),
+                    }),
+                },
+            }));
+        } catch (error) {
+            return this._processAdditionalAction(error, payment, billingAddress, orderAmount);
+        }
+    }
+
     private _mapToNonceInstrument(instrument: PaymentInstrument): NonceInstrument {
         return pick(instrument as NonceInstrument, 'nonce', 'shouldSaveInstrument', 'shouldSetAsDefaultInstrument');
```

Comment:
Can we use a typeguard here for instrument rather than asserting it using `as`

Reply:
I think we can